### PR TITLE
added oauth and reworked auth module

### LINKED ISSUE #208: Rework the role-model for the auth-module
### The problem faced currently?
The client can set the role when creating an user. This is a security issue, especially when the role is used for authorization rules in the config file.
### How can we solve it?
Rework the role-model so that no unauthorized and no unauthenticated entity can change the role of a user.
>If you want this feature to be implemented, give it a thumbs up reaction, so that we can determine which features are important to you.
>👍

### LINKED ISSUE #219: Add support for 0auth in user management module
Add support for 0Auth in user management module. We can start with the following:
- Twitter
- Github
- Google
- Facebook

>If you want this feature to be implemented, give it a thumbs up reaction, so that we can determine which features are important to you.
>👍

### PR DESCRIPTION
**Pull request by YourTechBud**
Fixes #208, fixes #219

### CODE REVIEW

File: modules/userman/oauth.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+		token, err := m.addUser(ctx, project, dbType, facebookUser.Email, facebookUser.Name)
+		if err != nil {
+			log.Println("Err: ", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
+			return
+		}
+
+		// Get the OAuth config
+		c, _ := m.getOAuth("google")
```
Comment (Inconnu08) [inline modules/userman/oauth.go]: Shouldn't this be `m.getOAuth("fb")`?

File: modules/userman/oauth.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+		token, err := m.addUser(ctx, project, dbType, githubUser.GetEmail(), githubUser.GetName())
+		if err != nil {
+			log.Println("Err: ", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
+			return
+		}
+
+		// Get the OAuth config
+		c, _ := m.getOAuth("google")
```
Comment (Inconnu08) [inline modules/userman/oauth.go]: Shouldn't this be `m.getOAuth("github")`?

File: modules/userman/oauth.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+		token, err := m.addUser(ctx, project, dbType, twitterUser.Email, twitterUser.Name)
+		if err != nil {
+			log.Println("Err: ", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			json.NewEncoder(w).Encode(map[string]string{"error": err.Error()})
+			return
+		}
+
+		// Get the OAuth config
+		c, _ := m.getOAuth("google")
```
Comment (Inconnu08) [inline modules/userman/oauth.go]: `m.getOAuth("twitter")`?

File: modules/userman/oauth.go (hunk 1 above)
Comment (YourTechBud) [reply modules/userman/oauth.go]: resolved

File: modules/userman/oauth.go (hunk 2 above)
Comment (YourTechBud) [reply modules/userman/oauth.go]: resolved

File: modules/userman/oauth.go (hunk 3 above)
Comment (YourTechBud) [reply modules/userman/oauth.go]: resolved

**Comment by YourTechBud** [comment]
@Inconnu08 I've fixed those issues. Do resolve those reviews if you think it's proper now.

**Comment by Inconnu08** [comment]
@YourTechBud LGTM 😊

File: modules/userman/oauth.go [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+		ctx := r.Context()
+		facebookUser, err := facebook.UserFromContext(ctx)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		token, err := m.addUser(ctx, project, dbType, facebookUser.Email, facebookUser.Name)
+		if err != nil {
```
Comment (Jayesh333) [inline modules/userman/oauth.go]: Why are we using different techniques for handling error?

File: modules/userman/oauth.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+	}
+	return http.HandlerFunc(fn)
+}
+
+func (m *Module) addUser(ctx context.Context, project, dbType, email, name string) (string, error) {
+	// Create a create request
+	readReq := &model.ReadRequest{Find: map[string]interface{}{"email": email}, Operation: utils.One}
+	_, err := m.crud.Read(ctx, dbType, project, "users", readReq)
+	if err == nil {
+		return "", errors.New("User with provided email already exists")
```
Comment (Jayesh333) [inline modules/userman/oauth.go]: Dont throw this error

File: modules/userman/oauth.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+	googleOAuth2 "golang.org/x/oauth2/google"
+)
+
+// OAuthConfig configures the main ServeMux.
+type OAuthConfig struct {
+	ClientID     string
+	ClientSecret string
+}
+
+func (m *Module) newGoogleOAuth(router *mux.Router) {
```
Comment (Jayesh333) [inline modules/userman/oauth.go]: These methods are not getting used anywhere

File: modules/userman/oauth.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+func (m *Module) addUser(ctx context.Context, project, dbType, email, name string) (string, error) {
+	// Create a create request
+	readReq := &model.ReadRequest{Find: map[string]interface{}{"email": email}, Operation: utils.One}
+	_, err := m.crud.Read(ctx, dbType, project, "users", readReq)
+	if err == nil {
+		return "", errors.New("User with provided email already exists")
+	}
+
+	// Create new user object
+	user := map[string]interface{}{"email": email, "name": name}
```
Comment (Jayesh333) [inline modules/userman/oauth.go]: No role here

File: config/config.go [hunk 8]
```diff
@@ -59,7 +59,28 @@ type Rule struct {
 }
 
 // Auth holds the mapping of the sign in method
-type Auth map[string]*AuthStub // The key here is the sign in method
+type Auth struct {
+	DefaultRole string `json:"defaultRole" yaml:"defaultRole"`
+	Email       *Email `json:"email" yaml:"email"`
+	Google      *OAuth `json:"google" yaml:"google"`
+	Facebook    *OAuth `json:"fb" yaml:"fb"`
```
[the remark this answers is not in the document]
Comment (YourTechBud) [reply config/config.go]: No

File: modules/userman/profile.go [hunk 9]
```diff
@@ -17,13 +17,6 @@ import (
 func (m *Module) HandleProfile() http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 
-		// Allow this feature only if the user management is enabled
```
[the remark this answers is not in the document]
Comment (YourTechBud) [reply modules/userman/profile.go]: added it back

File: config/config.go [hunk 10]
```diff
@@ -59,7 +59,28 @@ type Rule struct {
 }
 
 // Auth holds the mapping of the sign in method
-type Auth map[string]*AuthStub // The key here is the sign in method
```
[the remark this answers is not in the document]
Comment (YourTechBud) [reply config/config.go]: removed it

File: modules/userman/oauth.go (hunk 7 above)
Comment (YourTechBud) [reply modules/userman/oauth.go]: done

File: modules/userman/oauth.go (hunk 5 above)
Comment (YourTechBud) [reply modules/userman/oauth.go]: done

File: modules/userman/oauth.go (hunk 4 above)
Comment (YourTechBud) [reply modules/userman/oauth.go]: Just

**Comment by YourTechBud** [comment]
We wont be working on the user management module anymore

**Pull request closed without merging**